middleware/hooks.go

```diff
@@ -12,3 +15,135 @@ type Hook interface {
+		err = h.store.DeleteSeeder(req.InfoHash, req.Peer)
+		if err != nil && err != storage.ErrResourceDoesNotExist {
+			return ctx, err
+		}
+
+		err = h.store.DeleteLeecher(req.InfoHash, req.Peer)
+		if err != nil && err != storage.ErrResourceDoesNotExist {
+			return ctx, err
+		}
+	case req.Event == bittorrent.Completed:
+		err = h.store.GraduateLeecher(req.InfoHash, req.Peer)
+		return ctx, err
+	case req.Left == 0:
+		// Completed events will also have Left == 0, but by making this
+		// an extra case we can treat "old" seeders differently from
+		// graduating leechers. (Calling PutSeeder is probably faster
+		// than calling GraduateLeecher.)
+		err = h.store.PutSeeder(req.InfoHash, req.Peer)
+		return ctx, err
+	default:
+		err = h.store.PutLeecher(req.InfoHash, req.Peer)
+		return ctx, err
+	}
+
+	return ctx, nil
+}
+
+func (h *swarmInteractionHook) HandleScrape(ctx context.Context, _ *bittorrent.ScrapeRequest, _ *bittorrent.ScrapeResponse) (context.Context, error) {
+	// Scrapes have no effect on the swarm.
+	return ctx, nil
+}
+
+// ErrInvalidIP indicates an invalid IP for an Announce.
+var ErrInvalidIP = errors.New("invalid IP")
+
+// ErrMissingScrapeAddressLength indicates that the ScrapeAddressLengthKey
+// in the context for a Scrape was not set or not an int.
+var ErrMissingScrapeAddressLength = errors.New("missing address length for scrape")
+
+// ErrInvalidScrapeAddressLength indicates that the ScrapeAddressLength given
+// in the context for a scrape was invalid.
+var ErrInvalidScrapeAddressLength = errors.New("invalid scrape address length")
+
+type skipResponseHook struct{}
+
+// SkipResponseHookKey is a key for the context of an Announce or Scrape to
+// control whether the response middleware should run.
+// Any non-nil value set for this key will cause the response middleware to
+// skip.
+var SkipResponseHookKey = skipResponseHook{}
+
+type scrapeAddressLength struct{}
+
+// ScrapeAddressLengthKey is the key under which to store the length of the
+// requester's IP address for a scrape.
+// The value is expected to be of type int.
+var ScrapeAddressLengthKey = scrapeAddressLength{}
+
+type responseHook struct {
+	store storage.PeerStore
+}
+
+func (h *responseHook) HandleAnnounce(ctx context.Context, req *bittorrent.AnnounceRequest, resp *bittorrent.AnnounceResponse) (_ context.Context, err error) {
+	if ctx.Value(SkipResponseHookKey) != nil {
+		return ctx, nil
+	}
+
+	switch len(req.IP) {
+	case net.IPv4len:
+		s := h.store.ScrapeSwarm(req.InfoHash, false)
+		resp.Incomplete = s.Incomplete
+		resp.Complete = s.Complete
+		resp.IPv4Peers, err = h.store.AnnouncePeers(req.InfoHash, req.Left == 0, int(req.NumWant), req.Peer)
+	case net.IPv6len:
+		s := h.store.ScrapeSwarm(req.InfoHash, true)
+		resp.Incomplete = s.Incomplete
+		resp.Complete = s.Complete
+		resp.IPv6Peers, err = h.store.AnnouncePeers(req.InfoHash, req.Left == 0, int(req.NumWant), req.Peer)
+	default:
+		err = ErrInvalidIP
+	}
```

Comment:
This could probably be rewritten to not be a switch:

```go
if req.IP == nil {
  return errors.New("invalid IP")
}

s := h.store.ScrapeSwarm(req.InfoHash, req.IP == net.IPv6len)
resp.Incomplete = s.Incomplete
resp.Complete = s.Complete
resp.IPv4Peers, err = h.store.AnnouncePeers(req.InfoHash, req.Left == 0, int(req.NumWant), req.Peer)
```